stdlib/public/Concurrency/AsyncStreamBackpressure.swift

```diff
@@ -0,0 +1,1961 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+#if !SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY
+@available(SwiftStdlib 9999, *)
+public struct AsyncStreamAlreadyFinishedError: Error {}
+
+@available(SwiftStdlib 5.1, *)
+extension AsyncStream {
+    /// A mechanism to interface between producer code and an asynchronous stream.
+    ///
+    /// Use this source to provide elements to the stream by calling one of the `write` methods, then terminate the stream normally
+    /// by calling the `finish()` method.
+    @available(SwiftStdlib 9999, *)
+    public struct Source: Sendable {
+        /// A strategy that handles the backpressure of the asynchronous stream.
+        @available(SwiftStdlib 9999, *)
+        public struct BackPressureStrategy: Sendable {
+            /// When the high watermark is reached producers will be suspended. All producers will be resumed again once
+            /// the low watermark is reached.
+            @available(SwiftStdlib 9999, *)
+            public static func watermark(low: Int, high: Int) -> BackPressureStrategy {
+                BackPressureStrategy(internalBackPressureStrategy: .watermark(.init(low: low, high: high)))
+            }
+
+            private init(internalBackPressureStrategy: _InternalBackPressureStrategy) {
+                self._internalBackPressureStrategy = internalBackPressureStrategy
+            }
+
+            fileprivate let _internalBackPressureStrategy: _InternalBackPressureStrategy
+        }
+
+        /// A type that indicates the result of writing elements to the source.
+        @available(SwiftStdlib 9999, *)
+        @frozen
+        public enum WriteResult: Sendable {
+            /// A token that is returned when the asynchronous stream's backpressure strategy indicated that production should
+            /// be suspended. Use this token to enqueue a callback by  calling the ``enqueueCallback(_:)`` method.
+            public struct CallbackToken: Sendable {
+                let id: UInt
```

Comment:
it might be good to use a fixed with type UInt32/UInt64 explicitly to account for platform differences.